ChurchCRM/CRM
# Dashboard upgrade - Family Roles widget

#### What's this PR do?
This pull request rewrites some of the logic related to the Family Roles widget in PeopleDashboard.php:
- The **current design** only accommodates the default family roles present during an initial install.  Any custom family roles are only rendered as _"Other - {Gender}"_.
- The **updated design**, based on ORM, creates an array based on every gender-based permutation of every family role-based permutation.  This design only renders data in the widget when a demographic is greater than zero (so, for example, if there are no female children to be found, no data will be rendered for that demographic type).
- Additionally, a minor UI improvement was made by including a tooltip over the demographic percentage bar which displays the actual percentage value of the demographic.  Previously, a user would not have this information and could only "eyeball" the value.

#### What Issues does it Close?

Closes #3678.

#### Where should the reviewer start?
There are three major changes to the code:
- Since this upgrade utilizes ORM, includes to the relevant models were made.
- The `$demographicStats` variable was altered to utilize an ORM query.
- The `foreach ($demographicStats as $demStat)` construct was rewritten as follows:
    - Variables are declared for all males counted for a family role in the array, all females counted for a family role in the array, the family role ID, the family role name, and for all people counted in the entire database.
    - For each iteration of family roles in the array, a row will be written representing all male and female members of that cohort (if a gender is not present in that cohort, a row will not be written).

#### How should this be manually tested?
1. Execute the `git checkout lbridgman-Fix3678` and `git pull` commands.
1. Log into the ChurchCRM system.
1. In the header bar, choose _Your Tasks -> Other Settings -> Family Roles_ (/OptionManager.php?mode=famroles).
1. Add two new roles (for example, _Single Parent_ and _Houseguest_).
1. Choose _People -> Add New Person_.
1. Enter a name, a gender, and assign one of the new family roles created during this procedure.
1. Repeat the previous two steps as many time as you like.
1. Choose _People -> Dashboard_.
1. Verify that all the demographic permutations which exist in your database are represented.
1. Verify that demographic permutations which do not exist in your database are not represented.
1. Verify that the hyperlinks function as expected and produce an accurate number of entries.
1. Verify that the tooltip over the percentage bar renders properly and is accurate.

#### How should the automated tests treat this?
_With [sincere admiration](https://youtu.be/I0ncQup3BOE?t=199), Kirk._
#### Any background context you want to provide?
My church uses a lot of different family roles, such as young adults, single parents, singles, widows, houseguests, etc.  The Family Role widget, such as it was originally designed, didn't give us the insight we needed beyond the original default roles.  This redesign meets our requirements, lets me get a little more ORM experience, and also lets me render the percentage tooltip value while I'm at it.

#### What are the relevant tickets?

#### Screenshots (if appropriate)
**Before**
![image](https://user-images.githubusercontent.com/23409144/33802639-4d8086a0-dd49-11e7-8141-9c5be33081be.png)

**After**
![image](https://user-images.githubusercontent.com/23409144/33802649-856797f2-dd49-11e7-9866-da1f6fe6c1d1.png)
